go/runtime/history/db.go

```
@@ -214,6 +214,59 @@ func (d *DB) commit(blk *roothash.AnnotatedBlock, roundResults *roothash.RoundRe
 	})
 }
 
+func (d *DB) commitBatch(blks []*roothash.AnnotatedBlock, roundResults []*roothash.RoundResults) error {
+	if len(blks) != len(roundResults) {
+		return fmt.Errorf("runtime/history: got %d annotated blocks and %d roundResults, but should be equal", len(blks), len(roundResults))
+	}
+	if len(blks) == 0 {
+		return nil
+	}
+
+	return d.db.Update(func(tx *badger.Txn) error {
+		meta, err := d.queryGetMetadata(tx)
+		if err != nil {
+			return err
+		}
+
+		for i, blk := range blks {
+			rtID := blk.Block.Header.Namespace
+			if !rtID.Equal(&meta.RuntimeID) {
+				return fmt.Errorf("runtime/history: runtime mismatch (expected: %s got: %s)",
+					meta.RuntimeID,
+					rtID,
+				)
+			}
+
+			if blk.Height < meta.LastConsensusHeight {
+				return fmt.Errorf("runtime/history: commit at lower consensus height (current: %d wanted: %d)",
+					meta.LastConsensusHeight,
+					blk.Height,
+				)
+			}
+
+			if blk.Block.Header.Round <= meta.LastRound && meta.LastConsensusHeight != 0 {
+				return fmt.Errorf("runtime/history: commit at lower round (current: %d wanted: %d)",
+					meta.LastRound,
+					blk.Block.Header.Round,
+				)
+			}
+
+			if err := tx.Set(blockKeyFmt.Encode(blk.Block.Header.Round), cbor.Marshal(blk)); err != nil {
+				return err
+			}
+
+			if err := tx.Set(roundResultsKeyFmt.Encode(blk.Block.Header.Round), cbor.Marshal(roundResults[i])); err != nil {
+				return err
+			}
```

Might be a good idea to defensively check for nil pointers? Same applies for existing `Commit` method?

---

I am checking for it here now: https://github.com/oasisprotocol/oasis-core/blob/783354ef0305d8a34303a5b0b72e84d05c4661fe/go/runtime/history/db.go#L186

I think  from perspective of `runtime/history` being independent component this makes sense. Observe though that this changes existing behavior: previously if there was a critical bug in the upstream (roothash service sending nil blocks) you would panic here and reindex would stop. Now we only error, and technically since on error `reindexDone` is se to `false`,  we simply restart reindex on next event.

Any opinions? would it be better to return semantic error here and check for it in the roothash service + panic there. I have checked some previous sources, prior to #3091, where we panicked in such cases instead of continuously retrying.

Alternative is to remove this check and leave the code as it worked in the past and call this overkill.

---

https://github.com/oasisprotocol/oasis-core/pull/6050#discussion_r1961786105 resolved here.